migrator/migrations/m_111_to_m_112_groups_invalid_values/migration.go

```diff
@@ -0,0 +1,99 @@
+package m111tom112
+
+import (
+	"bytes"
+
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/migrator/migrations"
+	"github.com/stackrox/rox/migrator/types"
+	bolt "go.etcd.io/bbolt"
+)
+
+// groupStoredByCompositeKey is a helper struct which contains the group as well as the composite key.
+type groupStoredByCompositeKey struct {
+	grp          *storage.Group
+	compositeKey []byte
+}
+
+var (
+	bucketName = []byte("groups2")
+
+	emptyPropertiesCompositeKey = serializePropsKey(nil)
+
+	migration = types.Migration{
+		StartingSeqNum: 111,
+		VersionAfter:   &storage.Version{SeqNum: 112},
+		Run: func(databases *types.Databases) error {
+			return removeGroupsWithInvalidValues(databases.BoltDB)
+		},
+	}
+)
+
+func init() {
+	migrations.MustRegisterMigration(migration)
+}
+
+func removeGroupsWithInvalidValues(db *bolt.DB) error {
+	groupsWithInvalidValues, err := fetchGroupsToRemove(db)
+	if err != nil {
+		return errors.Wrap(err, "error fetching groups to remove")
+	}
+
+	if err := removeGroupsStoredByCompositeKey(db, groupsWithInvalidValues); err != nil {
+		return errors.Wrap(err, "error removing groups with invalid values")
+	}
+
+	return nil
+}
+
+func fetchGroupsToRemove(db *bolt.DB) (groupsStoredByCompositeKey []groupStoredByCompositeKey, err error) {
+	err = db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(bucketName)
+		// Pre-req: Migrating a non-existent bucket should not fail.
+		if bucket == nil {
+			return nil
+		}
+		return bucket.ForEach(func(k, v []byte) error {
+			// In a prior migration, the groups bucket was migrated from groups being stored by a composite key to groups
+			// being stored by a UUID.
+			// Within that migration, groups that had an empty role name were skipped during migration.
+			// This lead to bucket entries where the properties and role name was both empty, thus making it impossible
```

Comment:
Should we also delete groups with empty role names here as a pre-caution?

Reply:
We do delete groups with empty role names as long as they are still stored keyed by the composite key.

I don't think we need to remove groups with empty role names when they are stored using a unique ID - currently it would be possible to delete those via API directly (I also don't expect those to be there, but then again neither did I with the current situation). Still, I'd say let's not do this and add additional complexity to the migration.

Reply:
For me, empty role name = invalid group = should be deleted. This corresponds with the goal of this migration as I see it - fix our previous mistake and return the groups bucket to the valid state. Maybe I understood it wrong and we delete somewhere groups with an empty role, but definitely not during startup:
https://github.com/stackrox/stackrox/blob/6113b92dce180cd7856b1917569fb3fd18786037/central/group/datastore/singleton.go#L22-L27

I won't fixate on this though. If you feel strongly about not adding this to migration right now, let's put it on hold - for me, it makes sense to do it now due to our release cadence.